### ISSUE TITLE
Narrowing of list items does not work inside a loop that uses += on them

### ISSUE THREAD

**Issue by camillol** [microsoft/pyright#7381]
**Describe the bug**
This should type check cleanly:
```python
def loop_modify(maybe_strs: list[str | int]):
    while True:
        if isinstance(maybe_strs[-1], str):
            maybe_strs[-1] += "hello"  # here pyright thinks maybe_strs[-1] is str | int
```
Instead I get:
```
Operator "+=" not supported for types "str | int" and "Literal['hello']"
  Operator "+" not supported for types "int" and "Literal['hello']"Pylance[reportOperatorIssue](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportOperatorIssue)
```
Pyright 1.1.352 thinks the item is a `str | int`, even though `isinstance` should have narrowed it to `str`.

**Code or Screenshots**
See above. See also this set of examples:
```python
# GOOD: normally isinstance narrows list items
def one(maybe_strs: list[str | int]):
    if isinstance(maybe_strs[-1], str):
        maybe_strs += "hello"

# GOOD: it works inside a loop if you don't modify the list
def loop_no_modify(maybe_strs: list[str | int]):
    while True:
        if isinstance(maybe_strs[-1], str):
            maybe_strs[-1].upper()

# BAD: if I use +=, it doesn't narrow the type
def loop_modify(maybe_strs: list[str | int]):
    while True:
        if isinstance(maybe_strs[-1], str):
            maybe_strs[-1] += "hello"  # here pyright thinks maybe_strs[-1] is str | int

# GOOD: if I modify a different item, it understands it's a string
def loop_modify_other(maybe_strs: list[str | int]):
    while True:
        if isinstance(maybe_strs[-1], str):
            maybe_strs[0] = maybe_strs[-1] + "hello"
            x, y = maybe_strs[0], maybe_strs[-1]

# UGLY: if I first call upper, it narrows the type, and += becomes ok
def loop_read_then_modify(maybe_strs: list[str | int]):
    while True:
        if isinstance(maybe_strs[-1], str):
            maybe_strs[-1].upper()
            maybe_strs[-1] += "hello"
```

This looks like a reemergence of https://github.com/microsoft/pyright/issues/2983 or https://github.com/microsoft/pyright/issues/3236. The repro cases can hopefully help narrow it down.

**VS Code extension or command-line**
Checked this with pyright 1.1.352 on the command line. Same results in VSCode.

**Comment by erictraut** [comment]
Yes, this looks like it's related to evaluation ordering. If I expand the `+=` into its constituent operations, I'm able to repro the error some of the time and not others. I'll investigate further.

```python
def loop_modify(maybe_strs: list[str | int]):
    while True:
        if isinstance(maybe_strs[-1], str):
            x = maybe_strs[-1]
            y = x + "hello"
            maybe_strs[-1] = y
  ```

**Comment by erictraut** [comment]
This is included in pyright 1.1.353, which I just published.
